File: app/src/main/java/com/cyclecare/app/data/repository/DailyLogRepositoryImpl.kt
package com.cyclecare.app.data.repository

import com.cyclecare.app.data.local.dao.DailyLogDao
import com.cyclecare.app.data.local.entity.DailyLogEntity
import com.cyclecare.app.domain.model.CervicalMucusType
import com.cyclecare.app.domain.model.DailyLog
import com.cyclecare.app.domain.model.DischargeType
import com.cyclecare.app.domain.model.FlowIntensity
import com.cyclecare.app.domain.model.IntimacyType
import com.cyclecare.app.domain.model.Mood
import com.cyclecare.app.domain.model.Symptom
import com.cyclecare.app.domain.model.TestResult
import com.cyclecare.app.domain.repository.DailyLogRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.time.LocalDate
import javax.inject.Inject

class DailyLogRepositoryImpl @Inject constructor(
    private val dailyLogDao: DailyLogDao
) : DailyLogRepository {

    override fun getAllLogs(): Flow<List<DailyLog>> =
        dailyLogDao.getAllLogs().map { entities -> entities.map { it.toDomain() } }

    override suspend fun getAllLogsList(): List<DailyLog> =
        dailyLogDao.getAllLogsList().map { it.toDomain() }

    override suspend fun getLogByDate(date: LocalDate): DailyLog? =
        dailyLogDao.getLogByDate(date)?.toDomain()

    override suspend fun insertLog(log: DailyLog): Long =
        dailyLogDao.insertLog(log.toEntity())

    override suspend fun updateLog(log: DailyLog) =
        dailyLogDao.updateLog(log.toEntity())

    override suspend fun upsertLog(log: DailyLog): Long {
        val existing = dailyLogDao.getLogByDate(log.date)
        return if (existing == null) {
            insertLog(log)
        } else {
            updateLog(log.copy(id = existing.id))
            existing.id
        }
    }

    override suspend fun deleteLog(log: DailyLog) =
        dailyLogDao.deleteLog(log.toEntity())

    override suspend fun getLogsInRange(startDate: LocalDate, endDate: LocalDate): List<DailyLog> =
        dailyLogDao.getLogsInRange(startDate, endDate).map { it.toDomain() }

    private fun DailyLogEntity.toDomain() = DailyLog(
        id = id,
        date = date,
        flow = flow?.let { runCatching { FlowIntensity.valueOf(it) }.getOrNull() },
        mood = mood?.let { runCatching { Mood.valueOf(it) }.getOrNull() },
        symptoms = symptoms.mapNotNull { runCatching { Symptom.valueOf(it) }.getOrNull() },
        discharge = discharge?.let { runCatching { DischargeType.valueOf(it) }.getOrNull() },
        weightKg = weightKg,
        temperature = temperature,
        sleepHours = sleepHours,
        waterMl = waterMl,
        intimacy = runCatching { IntimacyType.valueOf(intimacy) }.getOrDefault(IntimacyType.NONE),
        ovulationTest = runCatching { TestResult.valueOf(ovulationTest) }.getOrDefault(TestResult.NOT_TAKEN),
        pregnancyTest = runCatching { TestResult.valueOf(pregnancyTest) }.getOrDefault(TestResult.NOT_TAKEN),
        cervicalMucus = cervicalMucus?.let { runCatching { CervicalMucusType.valueOf(it) }.getOrNull() },
        sexualActivity = sexualActivity,
        exerciseMinutes = exerciseMinutes,
        notes = notes
    )

    private fun DailyLog.toEntity() = DailyLogEntity(
        id = id,
        date = date,
        flow = flow?.name,
        mood = mood?.name,
        symptoms = symptoms.map { it.name },
        discharge = discharge?.name,
        weightKg = weightKg,
        temperature = temperature,
        ovulationTest = ovulationTest.name,
        pregnancyTest = pregnancyTest.name,
        intimacy = intimacy.name,
        waterMl = waterMl,
        cervicalMucus = cervicalMucus?.name,
        sexualActivity = sexualActivity,
        sleepHours = sleepHours,
        exerciseMinutes = exerciseMinutes,
        notes = notes
    )
}

File: cyclecare_flutter/lib/data/repositories/daily_log_repository_impl.dart
import 'dart:convert';
import 'package:drift/drift.dart' as drift;
import '../../domain/entities/daily_log.dart' as domain;
import '../../domain/repositories/daily_log_repository.dart';
import '../database/app_database.dart';

class DailyLogRepositoryImpl implements DailyLogRepository {
  final AppDatabase _database;

  DailyLogRepositoryImpl(this._database);

  @override
  Future<List<domain.DailyLog>> getAllDailyLogs() async {
    final logs = await _database.getAllDailyLogs();
    return logs.map(_toDomain).toList();
  }

  @override
  Future<domain.DailyLog?> getDailyLogByDate(DateTime date) async {
    final log = await _database.getDailyLogByDate(date);
    return log != null ? _toDomain(log) : null;
  }

  @override
  Stream<List<domain.DailyLog>> watchAllDailyLogs() {
    return _database.watchAllDailyLogs().map(
          (logs) => logs.map(_toDomain).toList(),
        );
  }

  @override
  Future<int> insertDailyLog(domain.DailyLog log) {
    return _database.insertDailyLog(
      DailyLogsCompanion.insert(
        date: log.date,
        flow: drift.Value(log.flow),
        mood: drift.Value(log.mood),
        symptoms: drift.Value(jsonEncode(log.symptoms)),
        discharge: drift.Value(log.discharge),
        weightKg: drift.Value(log.weightKg),
        temperature: drift.Value(log.temperature),
        ovulationTest: drift.Value(log.ovulationTest ?? ''),
        pregnancyTest: drift.Value(log.pregnancyTest ?? ''),
        intimacy: drift.Value(log.intimacy ?? false),
        waterMl: drift.Value(log.waterMl),
        cervicalMucus: drift.Value(log.cervicalMucus),
        sexualActivity: drift.Value(log.sexualActivity ?? false),
        sleepHours: drift.Value(log.sleepHours),
        exerciseMinutes: drift.Value(log.exerciseMinutes),
        notes: drift.Value(log.notes),
      ),
    );
  }

  @override
  Future<bool> updateDailyLog(domain.DailyLog log) {
    return _database.updateDailyLog(
      DailyLog(
        id: log.id,
        date: log.date,
        flow: log.flow,
        mood: log.mood,
        symptoms: jsonEncode(log.symptoms),
        discharge: log.discharge,
        weightKg: log.weightKg,
        temperature: log.temperature,
        ovulationTest: log.ovulationTest ?? '',
        pregnancyTest: log.pregnancyTest ?? '',
        intimacy: log.intimacy ?? false,
        waterMl: log.waterMl,
        cervicalMucus: log.cervicalMucus,
        sexualActivity: log.sexualActivity ?? false,
        sleepHours: log.sleepHours,
        exerciseMinutes: log.exerciseMinutes,
        notes: log.notes,
        createdAt: DateTime.now(),
      ),
    );
  }

  @override
  Future<int> deleteDailyLog(int id) {
    return _database.deleteDailyLog(id);
  }

  @override
  Future<List<domain.DailyLog>> getLogsInRange(
      DateTime start, DateTime end) async {
    final allLogs = await getAllDailyLogs();
    return allLogs
        .where((log) =>
            log.date.isAfter(start.subtract(const Duration(days: 1))) &&
            log.date.isBefore(end.add(const Duration(days: 1))))
        .toList();
  }

  domain.DailyLog _toDomain(DailyLog log) {
    List<String> symptoms = [];
    try {
      symptoms = (jsonDecode(log.symptoms) as List).cast<String>();
    } catch (e) {
      symptoms = [];
    }

    return domain.DailyLog(
      id: log.id,
      date: log.date,
      flow: log.flow,
      mood: log.mood,
      symptoms: symptoms,
      discharge: log.discharge,
      weightKg: log.weightKg,
      temperature: log.temperature,
      ovulationTest: log.ovulationTest.isEmpty ? null : log.ovulationTest,
      pregnancyTest: log.pregnancyTest.isEmpty ? null : log.pregnancyTest,
      intimacy: log.intimacy,
      waterMl: log.waterMl,
      cervicalMucus: log.cervicalMucus,
      sexualActivity: log.sexualActivity,
      sleepHours: log.sleepHours,
      exerciseMinutes: log.exerciseMinutes,
      notes: log.notes,
    );
  }
}
